# Тема 10. Декораторы и исключения.
Отчет по Теме #10 выполнила:
- Фадеева Анна Владимировна 
- АИС-22-1

| Задание | Лаб_раб | Сам_раб |
| ------ | ------ | ------ |
| Задание 1 | + | + |
| Задание 2 | + | + |
| Задание 3 | + | + |
| Задание 4 | + | + |
| Задание 5 | + | + |



знак "+" - задание выполнено; знак "-" - задание не выполнено;

Работу проверили:
- к.э.н., доцент Панов М.А.

## Лабораторная работа №10
### Задание 1. Наверняка вы думаете, что декораторы – это какая-то бесполезная вещь, которая вам никогда не пригодится, но тут вдруг на паре по математике преподаватель просит всех посчитать число Фибоначчи для 100. Кто-то будет считать вручную (так точно не нужно), кто-то посчитает на калькуляторе, а кто-то подумает, что он самый крутой и напишет рекурсивную программу на Python и немного огорчится, потому что данная программа будет достаточно долго считаться, если ее просто так запускать. Но именно тут к вам на помощь приходят декораторы, например @lru_cache (он предназначен для решения задач динамическим программированием, если простыми словами, то этот декоратор запоминает промежуточные результаты и при рекурсивном вызове функции программа не будет считать одни и те же значения, а просто “возьмёт их из этого декоратора”). Вам нужно написать программу, которая будет считать числа Фибоначчи для 100 и запустить ее без этого декоратора и с ним, посмотреть на разницу во времени решения поставленной задачи.

```python
from functools import lru_cache
@lru_cache(None)
def fibonacci(n):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    return fibonacci(n-1) + fibonacci(n-2)
if __name__ == '__main__':
    print(fibonacci(100))
```


### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20194835.png)
### Выводы.

1. декоратора @lru_cache:благодаря кэшированию повторяющихся вызовов функции, код работает быстрее для больших значений n. 

### Задание 2. Илья пишет свой сайт и ему необходимо сделать минимальную проверку ввода данных пользователя при регистрации. Для этого он реализовал функцию, которая выводит данные пользователя на экран и решил, что будет проверять правильность введённых данных при помощи декоратора, но в этом ему потребовалась ваша помощь. Напишите декоратор для функции, который будет принимать все параметры вызываемой функции (имя, возраст) и проверять чтобы возраст был больше 0 и меньше 130.Причем заметьте, что неважно сколько пользователь введет данных на сайт к Илье, будут обрабатываться только первые 2 аргумента.

```python
def check(input_func):
    def output_funk(*args):
        name, age = args[0], args[1]
        if age<0 or age>130:
            age = 'Недопустимый возраст'
        input_func(name, age)
    return output_funk
@check
def personal_info(name,age):
    print(f"Name: {name} Age: {age}")
if __name__ == '__main__':
    personal_info('Владимир', 38)
    personal_info('Александр', -5)
    personal_info('Пётр', 138, 15, 48, 2)
```

### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20195856.png)

### Выводы

1. Функция personal_info декорируется с помощью @check, основная функция (personal_info) остается неизменной, а проверка добавляется поверх нее.

### Задание 3. Вам понравилась идея Ильи с сайтом, и вы решили дальше работать вместе с ним. Но вот в вашем проекте появилась проблема, кто-то пытается сломать вашу функцию с получением данных для сайта. Эта функция работает только с данными integer, а какой-то недохакер пытается все сломать и вместо нужного типа данных отправляет string. Воспользуйтесь исключениями, чтобы неподходящий тип данных не ломал ваш сайт. Также дополнительно можете обернуть весь код функции в try/except/finally для того, чтобы программа вас оповестила о том, что выявлена какая-то ошибка или программа успешно выполнена.

```python
def data(*args):
    try:
        for i in range(len(*args)):
            try:
                result = (args[0][i] * 15)//10
                print(result)
            except Exception as ex:
                print (ex)
    except Exception as ex:
        print(ex)
    finally:
        print('Вся информация обработана')

if __name__ == '__main__':
    data([1, 15, 'Hello', 'i', 'try', 'to', 'crash', 'your', 'site', 38, 45])
```

### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20201004.png)
### Выводы

1. блоки try, except и finally играют ключевую роль в создании устойчивых к ошибкам и эффективных программ на Python, повышая общий уровень надежности и стабильности кода.

### Задание 4. Продолжая работу над сайтом, вы решили написать собственное исключение, которое будет вызываться в случае, если в функцию проверки имени при регистрации передана строка длиннее десяти символов, а если имя имеет допустимую длину, то в консоль выводиться “Успешная регистрация”


```python
def NegativeValueException(Exception):
    pass

def check_name(name):
    if len(name)>10:
        raise NegativeValueException('Длина более 10 символов')
    else:
        print('Успешная регистрация')
if __name__ =='__main__':
    name = '1234567890'
    check_name(name)
```
### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20202228.png)



### Выводы
1.  NegativeValueException - это пример создания специализированного исключения для определенной задачи. Если длина больше 10, вызывает пользовательское исключение NegativeValueException.


### Задание 5. После запуска сайта вы поняли, что вам необходимо добавить логгер, для отслеживания его работы. Готовыми вариантами вы не захотели пользоваться, и поэтому решили создать очень простую пародию. Для этого создали две функции: init() (вызывается при создании класса декоратора в программе) и call() (вызывается при вызове декоратора). Создайте необходимый вам декоратор. Выведите все логи в консоль.


```python
class SiteChecker:
    def __init__(self,func):
        print('> Класс SiteChecker метод __init__ успешный запуск')
        self.func = func
    def __call__(self):
        print('> Проверка перед запуском', self.func.__name__)
        self.func()
        print('Проверка безопасного включения')
@SiteChecker
def site():
    print('Усердная работа сайта')
if __name__ == '__main__':
    print('>> Сайт запущен')
    site()
    print('>> Сайт выклюючен')
```
### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20203255.png)


### Выводы

1.  SiteChecker реализован как класс, который сам себя вызывает при декорировании функции. Метод __init__ инициализирует объект декоратора перед его использованием. Метод __call__ определяет поведение объекта как вызываемой функции.





## Самостоятельная работа №10

### Задание 1. Вовочка решил заняться спортивным программированием на python, но для этого он должен знать за какое время выполняется его программа. Он решил, что для этого ему идеально подойдет декоратор для функции, который будет выяснять за какое время выполняется та или иная функция. Помогите Вовочке в его начинаниях и напишите такой декоратор.


```python
import time
def timing_decorator(func):
    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        execution_time = end_time - start_time
        print(f"\nФункция {func.__name__} выполнилась за {execution_time:.6f} секунд")
        return result
    return wrapper

@timing_decorator
def fad():
    fib1 = fib2 = 1
    for i in range(2, 200):
        fib1, fib2 = fib2, fib1 + fib2
        print(fib2, end='')

if __name__ == '__main__':
    fad()
```

### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20205010.png)


### Выводы

1. Декоратор `timing_decorator` создает вложенную функцию `wrapper`.
2. `Wrapper` измеряет время выполнения оригинальной функции `func`.
3. Функция fad() генерирует первые 200 чисел в ряде Фибоначчи.
4. timing_decorator используется для измерения времени выполнения функции fad()

### Задание 2. Посмотрев на Вовочку, вы также загорелись идеей спортивного программирования, начав тренировки вы узнали, что для решения некоторых задач необходимо считывать данные из файлов. Но через некоторое время вы столкнулись с проблемой что файлы бывают пустыми, и вы не получаете вводные данные для решения задачи. После этого вы решили не просто считывать данные из файла, а всю конструкцию оборачивать в исключения, чтобы избежать такой проблемы. Создайте пустой файл и файл, в котором есть какая-то информация. Напишите код программы. Если файл пустой, то, нужно вызвать исключение (“бросить исключение”) и вывести в консоль “файл пустой”, а если он не пустой, то вывести информацию из файла.


```python
def read_file(file_path):
    try:
        with open(file_path, 'r') as file:
            content = file.read()
            if not content:
                raise Exception("Файл пустой")
            else:
                print("Информация из файла:")
                print(content)
    except FileNotFoundError:
        print(f"Файл '{file_path}' не найден")
    except Exception as e:
        print(f"Исключение: {e}")

if __name__ == '__main__':
    # Пример использования для пустого файла
    print("Для пустого файла:")
    read_file('old.txt')

    print("\n" + "="*30 + "\n")

    # Пример использования для файла с информацией
    print("Для файла с информацией:")
    read_file('new.txt')
```

### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20211836.png)


### Выводы

1. Функция read_file:
1)Открывает файл в режиме чтения.
2)Читает весь содержимое файла.
3)Проверяет, есть ли содержимое.
4)Если файл пустой, выбрасывает исключение.В противном случае выводит содержимое файла.
2. FileNotFoundError: обрабатывает случай, когда файл не существует.
3. Ошибка при чтении файла 'new.txt' будет обработана в блоке except Exception as e, и сообщение об ошибке будет выведено.


### Задание 3. Напишите функцию, которая будет складывать 2 и введенное пользователем число, но если пользователь введет строку или другой неподходящий тип данных, то в консоль выведется ошибка “Неподходящий тип данных. Ожидалось число.”. Реализовать функционал программы необходимо через try/except и подобрать правильный тип исключения. Создавать собственное исключение нельзя. Проведите несколько тестов, в которых исключение вызывается и нет. Результатом выполнения задачи будет листинг кода и получившийся вывод в консоль.



```python
def add_two_and_number():
    try:
        user_input = input("Введите число: ")
        result = 2 + float(user_input)
        print(f"Результат сложения: {result}")
    except ValueError:
        print("Тип данных не подходит. ВВЕДИТЕ ЧИСЛО!!!")


if __name__ == '__main__':
    # Тест 1: норм число
    add_two_and_number()
    # Тест 2: строка вместо числа
    add_two_and_number()
    # Тест 3: другой тип данных
    add_two_and_number()
    # Тест 4: отрицательное число
    add_two_and_number()
```


### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20213132.png)


### Выводы

1. Функция add_two_and_number:
Запрашивает у пользователя ввод числа.
Пытается преобразовать введенное значение в число и добавить к нему 2.
Выводит результат.
2. Обработка исключений:
ValueError: обрабатывает случай, когда введенное значение не может быть преобразовано в число.


### Задание 4. Создайте собственный декоратор, который будет использоваться для двух любых вами придуманных функций. Декораторы, которые использовались ранее в работе нельзя воссоздавать. Результатом выполнения задачи будет: класс декоратора, две как-то связанными с ним функциями, скриншот консоли с выполненной программой и подробные комментарии, которые будут описывать работу вашего кода.


```python
import time


# Определяем класс декоратора
class LerinDecorator:
    def __init__(self, func):
        self.func = func

    def __call__(self, *args, **kwargs):
        start_time = time.time()  # время начала
        result = self.func(*args, **kwargs)
        end_time = time.time()  # время окончания
        execution_time = end_time - start_time  # время выполнения
        print(f"Время выполнения функции '{self.func.__name__}': {execution_time} секунд")  # Выводим время выполнения
        return result

    # Используем LerinDecorator для функции факториала


@LerinDecorator
def factorial(n):
    result = 1
    for i in range(1, n + 1):
        result *= i
    return result


# Используем LerinDecorator для функции генерации квадратов
@LerinDecorator
def generate_squares(n):
    return [i ** 2 for i in range(1, n + 1)]


# Основной блок выполнения
if __name__ == '__main__':
    print(factorial(17))  # Вычисляем и печатаем факториал числа 17
    print(generate_squares(17))  # Генерируем и печатаем квадраты чисел от 1 до 17
```


### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20215737.png)


### Выводы

1. Определение класса-декоратора LerinDecorator:
Использует методы __init__ и __call__.
Метод __call__ измеряет время выполнения декорируемой функции.
2. Декорирование функций:
Функция factorial(n) вычисляет факториал числа n. (Выводит время, затраченное на вычисление факториала числа 17. Также выводит результат)
Функция generate_squares(n) генерирует квадраты чисел от 1 до n. (Выводит время, затраченное на генерацию квадратов чисел от 1 до 17. Также выводит список квадратов чисел от 1 до 17.)



### Задание 5. Создайте собственный декоратор, который будет использоваться для двух любых вами придуманных функций. Декораторы, которые использовались ранее в работе нельзя воссоздавать. Результатом выполнения задачи будет: класс декоратора, две как-то связанными с ним функциями, скриншот консоли с выполненной программой и подробные комментарии, которые будут описывать работу вашего кода.


```python
import functools

# Базовый декоратор
def my_decorator(func):
    @functools.wraps(func)  # Сохраняем данные оригинальной функции
    def wrapper(*args, **kwargs):  # Вложенная функция
        print(f"Вызывается {func.__name__}")
        result = func(*args, **kwargs)
        print(f"{func.__name__} завершен")
        return result
    return wrapper

# Функция 1
@my_decorator
def hello(name):
    """Приветствует пользователя."""
    print(f"Привет, {name}!")

# Функция 2
@my_decorator
def sum(a, b):
    """Сумма двух чисел."""
    return a + b

# Тестирование функций
hello("Мир")
result = sum(7, 8)
print(result)
```

### Результат.
![Меню](https://github.com/kopcheniyferz/vAnnaofslobber/blob/тема-10/тема%2010/Снимок%20экрана%202024-11-17%20222207.png)


### Выводы

1. Определение базового декоратора my_decorator:
1)Использует вложенную функцию wrapper.
2)Сохраняет метаданные оригинальной функции с помощью functools.wraps.
2. Декорирование функций:
1)Функция hello(name) приветствует пользователя.
2)Функция sum(a, b) вычисляет сумму двух чисел.




## Общие выводы по теме
- На основе всех представленных выше заданий можно сделать выводы:

1. Декораторы в Python представляют собой мощный механизм, который позволяет расширять функциональность существующих функций. Этот подход позволяет разработчикам добавлять новое поведение к уже существующим функциям, не изменяя сам код этих функций.

2. Исключения играют критическую роль в обработке ошибок при работе с декораторами. Они служат эффективным инструментом для управления различными сценариями выполнения кода. В частности, исключения помогают справиться с ошибками, связанными с вводом/выводом данных, неправильными типами данных и другими потенциальными проблемами, которые могут возникнуть во время выполнения программы.

3. Этот подход к обработке ошибок особенно ценен при работе с декораторами, так как он позволяет разработчикам реагировать на различные ситуации, возникающие во время выполнения декорируемых функций, и обеспечивать стабильную работу программы даже в непредвиденных обстоятельствах.